### ISSUE TITLE
Facing linking error "undefined reference to `executorch::extension::FileDataLoader::from"

### ISSUE THREAD

**Issue by vikasbalaga** [pytorch/executorch#9636]
### 🐛 Describe the bug

Hi,

I am trying [Executorch C++ tutorial](https://pytorch.org/executorch/0.5/running-a-model-cpp-tutorial.html), by linking executorch libs into my custom application.

The compilation step is OK, but while linking my application, I am facing the below issue :

`undefined reference to executorch::extension::FileDataLoader::from(char const*, unsigned int)`

By following the tutorial, I have made the following changes (**I am showing only diff**) in my application :
```
/*Include header files*/
#include <executorch/extension/data_loader/file_data_loader.h>
#include <executorch/extension/memory_allocator/malloc_memory_allocator.h>
#include <executorch/extension/evalue_util/print_evalue.h>
#include <executorch/extension/runner_util/inputs.h>
#include <executorch/runtime/executor/method.h>
#include <executorch/runtime/executor/program.h>
#include <executorch/runtime/platform/log.h>
#include <executorch/runtime/platform/runtime.h>


using executorch::aten::Tensor;
using executorch::aten::TensorImpl;
using executorch::extension::FileDataLoader;
using executorch::extension::MallocMemoryAllocator;
using executorch::runtime::Error;
using executorch::runtime::EValue;
using executorch::runtime::HierarchicalAllocator;
using executorch::runtime::MemoryManager;
using executorch::runtime::Method;
using executorch::runtime::MethodMeta;
using executorch::runtime::Program;
using executorch::runtime::Result;
using executorch::runtime::Span;
```

```
/* main() */
	Result<FileDataLoader> loader =
			FileDataLoader::from("sample_model.pte");
	assert(loader.ok());

	Result<Program> program = Program::load(&loader.get());
	assert(program.ok());	
```

Also, I linked the following libs with my application :

```
libexecutorch.a
libexecutorch_core.a
libextension_runner_util.a
liboptimized_portable_kernels.a
libportable_kernels.a
libportable_ops_lib.a
libquantized_kernels.a
libquantized_ops_lib.a
```
I can confirm that, these libs don't contain any reference to `FileDataLoader` and I have seen some examples in the repo, which follow similar steps.

So, could you please help me in resolving this issue?

### Versions

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 10.0.0
CMake version: version 3.29.3
Libc version: glibc-2.35

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.35
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           11th Gen Intel(R) Core(TM) i5-11500 @ 2.70GHz
CPU family:                           6
Model:                                167
Thread(s) per core:                   2
Core(s) per socket:                   6
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          4600.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5424.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap avx512ifma clflushopt intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            288 KiB (6 instances)
L1i cache:                            192 KiB (6 instances)
L2 cache:                             3 MiB (6 instances)
L3 cache:                             12 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] numpydoc==1.5.0
[conda] _anaconda_depends         2023.09             py311_mkl_1
[conda] blas                      1.0                         mkl
[conda] mkl                       2023.1.0         h213fc3f_46343
[conda] mkl-service               2.4.0           py311h5eee18b_1
[conda] mkl_fft                   1.3.8           py311h5eee18b_0
[conda] mkl_random                1.2.4           py311hdb19cb5_0
[conda] numpy                     1.24.3          py311h08b1b3b_1
[conda] numpy-base                1.24.3          py311hf175353_1
[conda] numpydoc                  1.5.0           py311h06a4308_0
```

cc @larryliu0820 @jathu @lucylq

**Comment by vikasbalaga** [comment]
**UPDATE**
By referring to this [issue](https://github.com/pytorch/executorch/issues/7459#issuecomment-2573962872), I came to know that some compilation flags need to be turned ON, which I did as follows :

```
# Build
cmake                                                 \
    -DCMAKE_INSTALL_PREFIX=${et_build_dir}            \
    -DCMAKE_BUILD_TYPE=${build_type}                  \
    -DCMAKE_TOOLCHAIN_FILE="${toolchain_cmake}"       \
    -DEXECUTORCH_BUILD_EXECUTOR_RUNNER=OFF            \
    -DEXECUTORCH_BUILD_ARM_BAREMETAL=ON               \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON           \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON 	      \
    -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON       \
    -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON            \
    -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON            \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=OFF 	      \
    -DEXECUTORCH_BUILD_EXTENSION_RUNNER_UTIL=ON       \
    -DEXECUTORCH_ENABLE_LOGGING=ON                    \
    ${build_devtools_flags}                           \
    ${build_with_etdump_flags}                        \
    -B"${et_build_dir}"                               \
    "${et_root_dir}"
```
However, while building the "**extension_module**" lib, I am seeing the following compilation issue
```
executorch/extension/data_loader/mman.h:18:10: fatal error: sys/mman.h: No such file or directory
   18 | #include <sys/mman.h>
```

I couldn't find any such file in the `/sys` path of my machine.
Can someone please explain if I am missing something here?

**Comment by JacobSzwejbka** [comment]
Hmm not the most familiar, it seems like sys/mman should be on posix compliant systems. @tarun292  any idea on that? 

A different problem we could fix that would help you resolve this would be to split the data loaders into separate libraries in the cmake build. Then you could just not include the mmap data loader and avoid this dependency entirely on your system.

**Comment by vikasbalaga** [comment]
Hi @JacobSzwejbka,
Thanks for your quick response.
> A different problem we could fix that would help you resolve this would be to split the data loaders into separate libraries in the cmake build. Then you could just not include the mmap data loader and avoid this dependency entirely on your system.

I can try to implement this work around, until the fix is available. So, could you please elaborate more on the steps I need to take to implement it?

**Comment by tarun292** [comment]
@vikasbalaga have you tried 

```
sudo apt update
sudo apt install build-essential
sudo apt install libc6-dev
```

**Comment by vikasbalaga** [comment]
@JacobSzwejbka and @tarun292, thanks for your response.

I took a glance at the [arm_executor_runner ](https://github.com/pytorch/executorch/blob/main/examples/arm/executor_runner/arm_executor_runner.cpp)implementation and tried to replicate the same approach of compiling the model as a header file in my application.
With this I am able to build successfully! :)
But as I am trying to execute the application, it is crashing during some memory allocation at [method allocation](https://github.com/pytorch/executorch/blob/main/runtime/executor/tensor_parser_portable.cpp#L149). 
**I tried to debug if the issue is related to my heap memory, I can see the [tensor_impl](https://github.com/pytorch/executorch/blob/main/runtime/executor/tensor_parser_portable.cpp#L142) buffer is getting valid ptr address within expected memory range**
I am using a "**softmax**" model and allocating the following sizes of memory in my application
(**I hope this memory is sufficient for my basic example**)
```
#define ET_ARM_BAREMETAL_METHOD_ALLOCATOR_POOL_SIZE (512)
const size_t method_allocation_pool_size =
    ET_ARM_BAREMETAL_METHOD_ALLOCATOR_POOL_SIZE;
unsigned char __attribute__((
    aligned(16))) method_allocation_pool[method_allocation_pool_size];

#define ET_ARM_BAREMETAL_TEMP_ALLOCATOR_POOL_SIZE (512)
const size_t temp_allocation_pool_size =
    ET_ARM_BAREMETAL_TEMP_ALLOCATOR_POOL_SIZE;
unsigned char __attribute__((
    aligned(16))) temp_allocation_pool[temp_allocation_pool_size];
```
So, can you give me your insights, if I am missing anything here? (I am sure that sufficient heap memory is allocated for the app)

**Comment by vikasbalaga** [comment]
> [@JacobSzwejbka](https://github.com/JacobSzwejbka) and [@tarun292](https://github.com/tarun292), thanks for your response.
> 
> I took a glance at the [arm_executor_runner ](https://github.com/pytorch/executorch/blob/main/examples/arm/executor_runner/arm_executor_runner.cpp)implementation and tried to replicate the same approach of compiling the model as a header file in my application. With this I am able to build successfully! :) But as I am trying to execute the application, it is crashing during some memory allocation at [method allocator](https://github.com/pytorch/executorch/blob/main/runtime/executor/tensor_parser_portable.cpp#L149). **I tried to debug if the issue is related to my heap memory, I can see the [tensor_impl](https://github.com/pytorch/executorch/blob/main/runtime/executor/tensor_parser_portable.cpp#L142) buffer is getting valid ptr address within expected memory range** I am using a "**softmax**" model and allocating the following sizes of memory in my application (**I hope this memory is sufficient for my basic example**)
> 
> ```
> #define ET_ARM_BAREMETAL_METHOD_ALLOCATOR_POOL_SIZE (512)
> const size_t method_allocation_pool_size =
>     ET_ARM_BAREMETAL_METHOD_ALLOCATOR_POOL_SIZE;
> unsigned char __attribute__((
>     aligned(16))) method_allocation_pool[method_allocation_pool_size];
> 
> #define ET_ARM_BAREMETAL_TEMP_ALLOCATOR_POOL_SIZE (512)
> const size_t temp_allocation_pool_size =
>     ET_ARM_BAREMETAL_TEMP_ALLOCATOR_POOL_SIZE;
> unsigned char __attribute__((
>     aligned(16))) temp_allocation_pool[temp_allocation_pool_size];
> ```
> 
> So, can you give me your insights, if I am missing anything here? (I am sure that sufficient heap memory is allocated for the app)

Hi, again
After cleaning up my codebase, I can see this memory issue is not occurring :)

But now, I am seeing a different issue  `kernel 'aten::add.out' not found`

It looks like there is some operator registry in which all the ops need to be registered, but I am not sure how it works.
As I mentioned, my example is based on [arm_executor_runner](https://github.com/pytorch/executorch/blob/main/examples/arm/executor_runner/arm_executor_runner.cpp) and I coudn't see any OPs registering in that code

Also, I tried to follow the steps mentioned in this [Query](https://github.com/pytorch/executorch/issues/5009) by removing `libportable_ops_lib.a`, yet I am seeing the issue.

So could you please provide any info on these OPs registration procedure?

Thanks!

(Attacked logs below)

```
E [executorch:method.cpp:748] Missing operator: [0] aten::add.out
E [executorch:method.cpp:989] There are 1 instructions don't have corresponding operator registered.See logs for details
```

**Comment by JakubGonera** [comment]
I'm experiencing a similar issue, getting "Missing operator: [0] aten::cat.out" when building a C++ app on Android. I'm linking against these libraries:
`set(EXECUTORCH_LIBS 
  ${ET_LIB_DIR}/${ANDROID_ABI}/libcpuinfo.a                      
  ${ET_LIB_DIR}/${ANDROID_ABI}/libexecutorch_core.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libextension_data_loader.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libextension_module_static.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libextension_module.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libextension_runner_util.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libextension_tensor.a      
  ${ET_LIB_DIR}/${ANDROID_ABI}/libmicrokernels-prod.a      
  ${ET_LIB_DIR}/${ANDROID_ABI}/libpthreadpool.a
  ${ET_LIB_DIR}/${ANDROID_ABI}/libXNNPACK.a)

target_link_libraries(
    react-native-executorch
    ${EXECUTORCH_LIBS}
    "-Wl,--whole-archive" 
    ${ET_LIB_DIR}/${ANDROID_ABI}/libxnnpack_backend.a 
    ${ET_LIB_DIR}/${ANDROID_ABI}/libexecutorch.a
    ${ET_LIB_DIR}/${ANDROID_ABI}/libportable_kernels.a
    "-Wl,--no-whole-archive"
)`

What could be the issue here?

**Comment by JakubGonera** [comment]
I managed to solve the issue, in my case I was lacking linking libportable_ops_lib.a. Having this and libportable_kernels.a under --whole-archive made it work.
